pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableDebugResource.java

```diff
@@ -0,0 +1,267 @@
+   *
+   * @param tableNameWithType Name of table with type suffix
+   * @return Debug information for brokers of the table
+   */
+  private List<TableDebugInfo.BrokerDebugInfo> debugBrokers(String tableNameWithType) {
+    List<TableDebugInfo.BrokerDebugInfo> brokerDebugInfos = new ArrayList<>();
+    HelixDataAccessor helixDataAccessor = _pinotHelixResourceManager.getHelixZkManager().getHelixDataAccessor();
+    IdealState idealState =
+        helixDataAccessor.getProperty(helixDataAccessor.keyBuilder().idealStates(BROKER_RESOURCE_INSTANCE));
+    ExternalView externalView =
+        helixDataAccessor.getProperty(helixDataAccessor.keyBuilder().externalView(BROKER_RESOURCE_INSTANCE));
+
+    for (Map.Entry<String, String> entry : idealState.getInstanceStateMap(tableNameWithType).entrySet()) {
+      String brokerName = entry.getKey();
+      String isState = entry.getValue();
+      String evState = externalView.getStateMap(tableNameWithType).get(brokerName);
+      if (!isState.equals(evState)) {
+        brokerDebugInfos.add(
+            new TableDebugInfo.BrokerDebugInfo(brokerName, new TableDebugInfo.IsEvState(brokerName, isState, evState)));
+      }
+    }
+    return brokerDebugInfos;
+  }
+
+  /**
+   * Computes debug information for servers.
+   *
+   * @param pinotHelixResourceManager Helix Resource Manager
+   * @param tableName Name of table
+   * @param tableType Type of table (offline|realtime)
+   * @return Debug information for servers of the table
+   */
+  private List<TableDebugInfo.ServerDebugInfo> debugServers(PinotHelixResourceManager pinotHelixResourceManager,
+      String tableName, TableType tableType) {
+    HelixDataAccessor accessor = _pinotHelixResourceManager.getHelixZkManager().getHelixDataAccessor();
+    List<TableDebugInfo.ServerDebugInfo> serverDebugInfos = new ArrayList<>();
+
+    for (String instanceName : pinotHelixResourceManager.getServerInstancesForTable(tableName, tableType)) {
+      PropertyKey.Builder keyBuilder = accessor.keyBuilder();
+      List<String> sessionIds = accessor.getChildNames(keyBuilder.errors(instanceName));
+
+      if (sessionIds == null || sessionIds.size() == 0) {
+        return serverDebugInfos;
+      }
+
+      String tableNameWithType = TableNameBuilder.forType(tableType).tableNameWithType(tableName);
+      int numErrors = 0;
+      for (String sessionId : sessionIds) {
+        List<HelixProperty> childValues =
+            accessor.getChildValues(keyBuilder.errors(instanceName, sessionId, tableNameWithType), false);
+        numErrors = (childValues != null) ? childValues.size() : 0;
+      }
+
+      List<String> messageNames = accessor.getChildNames(accessor.keyBuilder().messages(instanceName));
+      int numNewMessages = 0;
+      int numReadMessages = 0;
+      if (messageNames != null) {
+        for (String messageName : messageNames) {
+          Message message = accessor.getProperty(accessor.keyBuilder().message(instanceName, messageName));
```

Comment:
This might be very costly, especially when there are lots of messages stuck in the queue. Maybe just log the `numMessages`?

Reply:
Yeah, however, we won't have NEW/READ messages. I think that should be ok, we definitely don't want to read lots of unstuck messages.